Add an ABI fixup for generic context

Kira-NT opened the pull request:
### Description

Currently, if you attempt to inject into a generic method while using .NET, the injection itself works, but the hook often receives incorrect arguments. This issue arises because .NET passes a generic context to some generic methods in the form of `MethodDesc`/`MethodTable`/`TypeHandle`, which shifts all user arguments by one position.

`MonoMod` already supports the concept of "ABI fixups," so I resolved the issue by implementing a new ABI fixup rule for these scenarios. This rule disregards the provided context, similarly to how it's currently done in the Mono realm. In Mono, the `r10` register is used instead of the hidden argument for this purpose, so it doesn't create such a problem, and we can implicitly ignore its existence.

While this fix does not alter the official stance that "MonoMod doesn't support generics," it allows generics to function in .NET under conditions where they already somewhat work in Mono. Thus, this can be considered a parity feature :)

### Examples

Starting with .NET 5, all the following methods are now injectable:

```csharp
public class Foo
{
    public static void StaticMethod<T>(T t);

    public void InstanceMethod<T>(T t);
}

public class Bar<T>
{
    public static void StaticMethod(T t);

    public void InstanceMethod(T t);

    public static void StaticMethod2<U>(T t, U u);

    public void InstanceMethod2<U>(T t, U u);
}

public struct Baz
{
    public static void StaticMethod<T>(T t);

    public void InstanceMethod<T>(T t);
}

public struct Qux<T>
{
    public static void StaticMethod(T t);

    public void InstanceMethod(T t);

    public static void StaticMethod2<U>(T t, U u);

    public void InstanceMethod2<U>(T t, U u);
}
```

However, since this PR does not aim to address deeper issues, some existing problems still persist. For instance, certain injections fail to work on Windows *(though they function perfectly on Linux and macOS)* because MonoMod silently fails to apply them. Here are some scenarios I have identified:

```csharp
public void InstanceVoid<T>(List<T> list, T item, ref bool injected);

public string InstanceRegister<T>(List<T> list, T item, ref bool injected);

public LargeStruct InstanceByRef<T>(List<T> list, T item, ref bool injected);

public static LargeStruct StaticByRef<T>(List<T> list, T item, ref bool injected);
```

Thankfully, most common use cases "just work" now. And honestly, it's super useful when they do!

### Notes

Fixes #185

Kira-NT committed:
Refactor ABI fixup generation logic

Currently, it is hardcoded to support only a very specific scenario where an instance method that returns a value by reference is being detoured to a static method.

However, with this rewrite in mind, it should be much easier to introduce new conditions for generating ABI fixups in the future.

Additionally, I have added [quite] a bit more information on why and when these fixups are needed in the first place.

This change does **not** alter the existing behavior in **any** way.

Kira-NT committed:
Add condition to create fixups for generic methods

Additionally, introduce a missing processing branch for the `GenericContext`.
Without it, the generic context is treated as a user argument, which is just wrong.

Fixes #185

nike4613 commented on src/MonoMod.Core/Platforms/PlatformTriple.cs:
```diff
@@ -626,106 +626,234 @@ public MethodBase GetRealDetourTarget(MethodBase from, MethodBase to)
+                        // While this does work fine, it introduces funny little
+                        // holes in the type-safety premise of .NET:
+                        //
+                        // // This may return `false`!
+                        // bool Hook<T>(T it)
+                        //     => typeof(T).IsAssignableFrom(it.GetType());
+                        //
+                        // This behavior is caused by the fact that constructed
+                        // generics for reference types are Javad into a single
+                        // definition at runtime (and rightfully so).
+                        // So, even if you try to detour a specific generic
+                        // implementation using something like
+                        // `to.MakeGenericMethod(typeof(string))`,
+                        // your hook will receive calls for all
+                        // reference type-based implementations out there.
+                        // However, since we do not patch the generic context
+                        // of the provided hook, it remains unchanged,
+                        // causing `typeof(T)` to defy users' expectations.
+                        //
+                        // So, TODO: patch the generic context of the detour target
+                        // and/or introduce a call filter based on the current context.
```
This change in behavior (towards silently not doing what the user wants) drives me to want to also block generics in the RuntimeDetour API (though I'm OK with leaving it open from the Core API).

Kira-NT replied:
This makes sense, especially to avoid new issues like, *"Generic injections don't work the way I want!"* Where would you prefer to place this roadblock? I believe `Hook.PrepareRealTarget` could be a good candidate for that, as it already handles signature checks and similar stuff.

nike4613 replied:
That's probably where I'd put it, yes.

nike4613 commented on src/MonoMod.Core/Platforms/PlatformTriple.cs:
```diff
@@ -626,106 +626,234 @@ public MethodBase GetRealDetourTarget(MethodBase from, MethodBase to)
+                return false;
+
+            // If the method itself is generic, check whether at least one
+            // of its generic arguments makes it a shared instantiation.
+            var methodGenericArguments = method.IsGenericMethod ? method.GetGenericArguments() : Type.EmptyTypes;
+            if (methodGenericArguments.Any(static x => !x.IsValueType))
+                return true;
 
-            return to;
+            // If the method is effectively generic (i.e., it is defined on a generic type),
+            // we need to determine whether it still requires a hidden argument in the form
+            // of a MethodDesc, MethodTable, or TypeHandle pointer, or if `this` alone will
+            // suffice for this purpose.
+            //
+            // The rules are as follows:
+            //  - If both the method and its declaring type are generic, `this` alone cannot
+            //    provide the generic context the VM needs. In such cases,
+            //    we require a hidden argument for shared instantiations.
+            //  - If the method is static, there is obviously no `this` from which
+            //    the VM could infer the generic context.
+            //  - If `this` is a value type, the unboxed `this` pointer, by definition,
+            //    does not have an associated MethodTable pointer.
+            //  - If the method is a default interface method called via interface dispatch,
+            //    the VM cannot use `this` to derive the generic context because it is too
+            //    ambiguous (e.g., it may implement multiple IFoo<T> interfaces).
+            //
+            // See:
+            // https://github.com/dotnet/runtime/blob/55eee324653e01cf28809d02b25a5b0894b58d22/src/coreclr/vm/method.cpp#L1649
+            var mayNeedHiddenArg = method.IsGenericMethod
+                || method.IsStatic
+                || declaringType.IsValueType
+                || declaringType.IsInterface && !method.IsAbstract;
+            if (!mayNeedHiddenArg)
+                return false;
+
+            // Finally, check whether the type on which the method is defined may be shared.
+            var typeGenericArguments = declaringType.IsGenericType ? declaringType.GetGenericArguments() : Type.EmptyTypes;
+            return typeGenericArguments.Any(static x => !x.IsValueType);
```
All of this logic is runtime-specific, and so should call some runtime-specific helper to make this determination. The rules implemented here are also wrong: it mishandles the case of `M<SomeStruct<string>>()` *which is still shared*, even though the type argument is a value-type. (I'll point you over [here](https://github.com/nike4613/MonoMod/tree/coreclr-generics) and [here](https://github.com/MonoMod/MonoMod.Common/tree/15d444c207a49232318f423392177f0dbea82548/RuntimeDetour/Platforms/Generic) for something to reference-- I'm pretty sure I got the logic fully correct there.)

Kira-NT replied:
> All of this logic is runtime-specific

Yup, I initially wanted to put it in `IRuntime`, but that would be a breaking change since it impacts a public API. So, instead, I decided to place the method here to avoid a more significant change, as it realistically only affects CoreCLR at the moment :)

If you're okay with `IRuntime.RequiresGenericContext(MethodBase)` *(or something like that)*, I'll be happy to fix it and move the method to `Core[...]Runtime` while creating stubs for the other runtimes.  

> it mishandles the case of `M<SomeStruct<string>>()`

Oh, right - great catch! I basically followed the same CLR logic you referenced in the comments of your code but made the mistake of trivializing shared generic types to just non-value types, while somehow forgetting in the moment that value types can also be generic and, thus, shared :D

nike4613 replied:
The way I'm treating Core versioning is that any non-*consumer* breaking change is permitted across minor version bumps, and that certain consumer-breaking changes are permitted as long as they're not at the main entrypoint (i.e. DetourFactory). If we really don't like adding a method to `IRuntime`, then we can also add another interface to the implementers of `IRuntime`, like I did for CET, but I'd prefer just changing `IRuntime`.

nike4613 reviewed (changes requested):
I'm interested in merging this, if only so that we can have this part ready when we *do* actually get around to generics. It does need some work though, and (as I mentioned in one of the other comments) I would want to block generics in RuntimeDetour with this, precisely because the effects of trying to patch generics are less catastrophic.

Kira-NT committed:
Forbid hooking into generic methods

This scenario is currently not handled properly. So, to avoid potential confusion for users, we explicitly prohibit hooking into generic methods in the RuntimeDetour API.

Kira-NT committed:
Implement `IRuntime.RequiresGenericContext`

Kira-NT committed:
Update suppressions

nike4613 commented on src/MonoMod.Core/Platforms/Runtimes/FxCoreBaseRuntime.cs:
```diff
@@ -161,6 +162,62 @@ public virtual RuntimeMethodHandle GetMethodHandle(MethodBase method)
+            return typeGenericArguments.Any(IsGenericSharedType);
+        }
+
+        private static bool IsGenericSharedType(Type type)
+        {
+            // Primitives are never shared.
+            // Value types are shared only if at least one of their nested generic arguments is shared.
+            // All other types (i.e., reference types) are always shared.
+            return (
+                !type.IsPrimitive &&
+                (!type.IsValueType || type.IsGenericType && type.GetGenericArguments().Any(IsGenericSharedType))
+            );
+        }
```
It might be worth caching this somewhere, possibly in a `CWT<Type, StrongBox<bool>>`? Esp. seeing as it can recurse arbitrarily deeply. Not sure how important that is though.

Kira-NT replied:
While it *theoretically* can recurse quite deeply, I don't think it's worth optimizing for this scenario in practice. For this method to become substantially computationally expensive, we would need to encounter a deeply nested generic value type in the form of `Struct0<Struct1<Struct2<Struct3<...>, ...>, ...>, ...>` with no reference-type generic arguments to break the loop. And even then, how deep would it really need to be to cause a problem? Even a hundred levels of nesting would be processed in an instant; and in all the years I've worked with .NET, the worst atrocities I've encountered haven't gone beyond "just" a few dozens of total generic arguments.

This is such an insanely rare *(if not outright unrealistic)* scenario that, to be quite frankly honest with you, if someone does create types like this, they deserve to face the consequences of their actions! :D

Given this, I feel like polluting the runtime with a permanent cache for something we might not even encounter a second time would be a significantly worse tradeoff than simply computing the result of this relatively cheap operation as is.

nike4613 replied:
I have written such code, though in general I do agree.
